Added Named Accessors to RigidBodyPlant's Ports and Made Plant-Centric Actuator Command Input Port Only Available in Mono-Model-instance RigidBodyPlants.

liangfok opened the pull request:
This will be more robust against future `RigidBodyPlant` port number changes due to the addition of ports.

The actuator command input ports are modified as follows:

1. There is only one model instance-centric actuator command port per model instance with an actuator. (It is now possible for the plant to not have any input ports when there are no actuators.)

2. The plant-centric actuator command port is only available when there is only one model instance in the RigidBodyTree, and it returns the same port as the model-instance-centric port using the first model instance ID.

Resolves #4025.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4766)
<!-- Reviewable:end -->

liangfok commented:
+@seancurtis-tri for feature review please since I removed one of your `TODO` comments.
+@jwnimmer-tri for platform review please.

Thanks!

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaIy_DzD57E2fsA2pPJ:b-s1eeg3)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Please re-add me (or whoever) for review once dynamic team feature review is complete.
-@jwnimmer-tri 

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaJGwc0DwK_RM1GZsQK:b56yxnb)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few editorial comments.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJWfZsmiKopFIiPDXV:-KaJWfZsmiKopFIiPDXW:b-u0xhdd) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L37)):*
> ```cpp
>   DRAKE_DEMAND(tree_ != nullptr);
> 
>   // Declares an input port for the generalized force vector. The generalized
> ```

BTW I find it interesting that this vector is defined in two ways: "generalized forces" and "command".  But the two concepts don't seem to be explicitly connected in the documentation.  In fact, this appears to be the only place that suggests "commands" = "forces".  So, either this should just say it's the "command vector", or the connection between the two words should be made somewhere.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJXYVPWRc43J6X878Q:-KaJXYVPWRc43J6X878R:b-iql8mi) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L72)):*
> ```objc
> ///
> /// - model_state_output_port(): A vector-valued port containing the state
> ///   vector for a particular model instance within the RigidBodyTree that is
> ```

BTW this seems overly flowery.  (within...within).  How about, "...for a particular model defined in this RigidBodyPlant's RigidBodyTree?"

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJXvbX5ScwenHCVZgA:-KaJXvbX5ScwenHCVZgB:bjvlcxq) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L252)):*
> ```objc
>   /// These are accessors for obtaining descriptors of this
>   /// RigidBodyPlant's input port. See this class's description for details
>   /// about this port.
> ```

BTW Plurality issues?   i.e., previous doxygen says there's a single input, but this comment has the words "accessors" and "descriptors".  But then it says "input port" and "this port".  Copy and paste, I assume.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 299 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJYYnkpZyikfIYcM-P:-KaJYYnlbYVtHrOiCzyi:bhdkqr9) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L299)):*
> ```objc
>   /// @name System output port index accessors.
>   /// These are accessors for obtaining indices of this RigidBodyPlant's output
>   /// ports. See this class's description for details about these ports.
> ```

BTW I'd love some hint of when these would be used (as part of a public api.)  That's one of those vague aesthetics of mine that I like, but don't consistently conform to.  But in the light of the earlier documentation where we can grab the port descriptors directly, the existence of these seem could use some justification in the docs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaJZ31w34C169gzecpb:b-1k2c0h)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJWfZsmiKopFIiPDXV:-KaJq6051JONWH9j-B91:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L37)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I find it interesting that this vector is defined in two ways: "generalized forces" and "command".  But the two concepts don't seem to be explicitly connected in the documentation.  In fact, this appears to be the only place that suggests "commands" = "forces".  So, either this should just say it's the "command vector", or the connection between the two words should be made somewhere.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJXYVPWRc43J6X878Q:-KaJqy_xbq517gFSW5lc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L72)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW this seems overly flowery.  (within...within).  How about, "...for a particular model defined in this RigidBodyPlant's RigidBodyTree?"
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJXvbX5ScwenHCVZgA:-KaJrmBPxARanhII1-VJ:buqvxxx) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L252)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Plurality issues?   i.e., previous doxygen says there's a single input, but this comment has the words "accessors" and "descriptors".  But then it says "input port" and "this port".  Copy and paste, I assume.
</blockquote></details>

Done. It turns out `RigidBodyPlant` has more than one input port. I've colocated the input port accessors and updated the documentation.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 299 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaJYYnkpZyikfIYcM-P:-KaJvu2fa3SDs8jzeU-u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b0c6d483b2e5dada2aa418f070b3b88cbb7781ab/drake/multibody/rigid_body_plant/rigid_body_plant.h#L299)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I'd love some hint of when these would be used (as part of a public api.)  That's one of those vague aesthetics of mine that I like, but don't consistently conform to.  But in the light of the earlier documentation where we can grab the port descriptors directly, the existence of these seem could use some justification in the docs.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@seancurtis-tri: Thanks for the review. I've addressed your comments. PTAL.
+@jwnimmer-tri for platform review please. Thanks!

---

Review status: 5 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaJw4p73GyW5Q8WYYsj:bo7mr6j)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
As long as it all doxygenates nicely, I'm good.

---

Reviewed 2 of 2 files at r2.
Review status: 6 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaJyBVsn98kCX4iMum7:b-6dbqd6)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
-@jwnimmer-tri +@sherm1 for platform review, please.

---

Review status: 6 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaK7NKCPBChU3nyIyB7:bkms3oy)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHC0igBVmUPeBOLfX:-KaNHC0igBVmUPeBOLfY:bacmxig) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L35)):*
> ```objc
> ///   by model_input_port() can be connected. The size of this vector is equal
> ///   to the number of RigidBodyActuator's in the RigidBodyTree. Each
> ///   RigidBodyActuator maps to a single DOF joint (currently actuation cannot
> ```

BTW THe grammar nazi in me likes hyphens.  E.g.,: 
   "...a single DOF joint..." becomes "...a single-DOF joint..."
In the next line:
   "...applied to multiple DOF's joints." becomes "...applied to multiple-DOF joints."  (Note additional change from DOF's to DOF.)

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 47 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHe-BBHW24-2DRbWj:-KaNHe-BBHW24-2DRbWk:b4rgkot) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L47)):*
> ```objc
> ///   If any port returned by model_input_port() is connected, the input port
> ///   for the full tree, which is obtained using command_input_port(), must not
> ///   be used.
> ```

It seems to me that between this "rule" and its dual above, it is worth noting that this exclusive connections are not enforced by the system and that breaking this "rule" will lead to...undefined behavior?  Some such thing.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 77 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHzKaqWiwZe7tTunB:-KaNHzKaqWiwZe7tTunC:bn2ucq3) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L77)):*
> ```objc
> /// is applied to a RigidBodyTree, which is a multibody model that consists of a
> /// set of rigid bodies connected through joints in a tree structure. Bodies may
> /// have a collision model in which case collisions are considered. In addition,
> ```

"...model, in which case, collisions..."

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 78 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNI4CAI8dfk9_W8Rax:-KaNI4CAI8dfk9_W8Ray:b-7mxu2f) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L78)):*
> ```objc
> /// set of rigid bodies connected through joints in a tree structure. Bodies may
> /// have a collision model in which case collisions are considered. In addition,
> /// the model may contain loop constraints described by RigidBodyLoop's in the
> ```

"RigidBodyLoop's": apostrophe-s implies possession or contraction with "is".  The most "correct" may be RigidBodyLoop instances (which feels verbose to me, but it's usually the fall back I go to when I'm unable to come up with anything equally grammatical.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Partial address of reviewer comments. Trying to figure out what exactly happens when both types of RigidBodyPlant's input ports are hooked up. @sammy-tri do you know?

---

Review status: 5 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHC0igBVmUPeBOLfX:-KaNQfFLUhyrvuS-FzfD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L35)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW THe grammar nazi in me likes hyphens.  E.g.,: 
   "...a single DOF joint..." becomes "...a single-DOF joint..."
In the next line:
   "...applied to multiple DOF's joints." becomes "...applied to multiple-DOF joints."  (Note additional change from DOF's to DOF.)
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 77 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHzKaqWiwZe7tTunB:-KaNTA4lo88LNSFsVBRn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L77)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

"...model, in which case, collisions..."
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 78 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNI4CAI8dfk9_W8Rax:-KaNTJuFaI3fPRNoHe6Q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L78)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

"RigidBodyLoop's": apostrophe-s implies possession or contraction with "is".  The most "correct" may be RigidBodyLoop instances (which feels verbose to me, but it's usually the fall back I go to when I'm unable to come up with anything equally grammatical.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaOOGbKntMhsnM52k5M:b597mss)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I look forward to finding out how those ports are handled.  :)

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaOU3BWgFriKY9MCTZQ:b-bo7tqs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 302 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaLOiGM-TnudQ6iLoP3:-KaLOiGM-TnudQ6iLoP4:bn7x237) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L302)):*
> ```objc
>   int state_output_port_index() const {
>     return state_output_port().get_index();
>   }
> ```

These four methods seem like pointless verbiage to me. What is wrong with `port.get_index()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 4 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 302 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaLOiGM-TnudQ6iLoP3:-KaOoxJIUsUfOKvrss7e:bxdc7fw) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L302)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

These four methods seem like pointless verbiage to me. What is wrong with `port.get_index()`?
</blockquote></details>

Done. (Removed.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform pass done, PTAL.

---

Reviewed 4 of 7 files at r1, 1 of 2 files at r2, 2 of 2 files at r5.
Review status: 4 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOqRiyoEUu_NO9oa2I:-KaOqRiyoEUu_NO9oa2J:bct44nf) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L31)):*
> ```objc
> /// <B>Input Port Accessors:</B>
> ///
> /// - command_input_port(): Contains the command vector for the RigidBodyTree's
> ```

If this is an actuator input it should have `actuator` in the name. There could be lots of different types of commands I think.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 33 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOmRaJH-UVQp9xkRO1:-KaOmRaJH-UVQp9xkRO2:b7wxkp3) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L33)):*
> ```objc
> /// - command_input_port(): Contains the command vector for the RigidBodyTree's
> ///   actuators. Note that if this port is connected, none of the ports returned
> ///   by model_input_port() can be connected. The size of this vector is equal
> ```

This should not have both input ports if they can't both be used. Instead, specify on construction (or post-construction set methods) so that only one of the two input ports is created.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 71 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOpte1JGvPkKY4IyRT:-KaOpte1JGvPkKY4IyRU:b-cid796) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L71)):*
> ```objc
> ///
> /// - model_state_output_port(): A vector-valued port containing the state
> ///   vector for a particular model instance in the RigidBodyTree.
> ```

BTW, consider having both input and output obey the construction option. So you either get `all_actuators_in->RBPlant->all_states_out` or `model_instance_actuators_in->RBPlant->mode_instance_states_out`.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 312 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOln6lMDHmu0JhMy1B:-KaOln6lMDHmu0JhMy1C:b-b3up9u) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L312)):*
> ```cpp
>   // Evaluates the correctness of the kinematics results port.
>   auto& kinematics_results =
>       output_->get_data(kuka_plant_->kinematics_results_output_port_index())->
> ```

BTW, use get_index() as elsewhere in this file.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 533 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOlNSomDNz1J2YkjXc:-KaOlNSomDNz1J2YkjXd:b-7r4xhs) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L533)):*
> ```cpp
>       AddModelInstancesFromSdfFile(drake::GetDrakePath() +
>           "/multibody/rigid_body_plant/test/limited_prismatic.sdf",
>           kQuaternion, nullptr /* weld to frame */, tree.get());
> ```

BTW, is that comment right? I don't see a weld.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaOqnwy0Z79J9IFAi_c:b-fzvf1a)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 533 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOlNSomDNz1J2YkjXc:-KaOsZq-flhCnZ8R4GlC:bht00kn) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L533)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, is that comment right? I don't see a weld.
</blockquote></details>

I believe the comment represents how the `nullptr` is to be interpreted: as the `weld_to_frame` argument.  It appears in Drake frequently and *very* often in this file. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Great idea about forcing the ports of RigidBodyPlant to either be whole-plant-centric or model-instance-centric. I will adjust this PR accordingly. It will result in a larger PR, however.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 533 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOlNSomDNz1J2YkjXc:-KaP0V5MuJGbXswXD6AF:b-xnje62) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L533)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I believe the comment represents how the `nullptr` is to be interpreted: as the `weld_to_frame` argument.  It appears in Drake frequently and *very* often in this file. 
</blockquote></details>

@seancurtis-tri is right, the `nullptr` _is_ the `weld_to_frame` argument. I believe this is technical debt that will be paid off once #3093 and #4407 are resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaP1MKevF_itsuqd-3H:bdh11ru)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
> I will adjust this PR accordingly

Please see if you can make it so the current behavior remains the default to minimize changes.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaPALLLTsXc_WhgKbak:b6hxerl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I am in the midst of modifying this PR to XOR the provisioning of plant-centric vs. model-instance-centric ports and just had a revelation: how about only providing plant-centric ports when there is only one model instance in the `RigidBodyTree`? There's really no good reason why `RigidBodyTree` users will want multiple model instances to be lumped together into a single inport / output port. When there is only one model instance in the `RigidBodyTree`, the plant-centric ports can simply call the model-instance-centric port using model instance ID zero (or one once #3088 is resolved). I plan on pivoting in this new direction unless someone tells me I'm going in the wrong direction.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 312 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOln6lMDHmu0JhMy1B:-KaPGJRcOrmoGxNa69pi:b-wyw1fr) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L312)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, use get_index() as elsewhere in this file.
</blockquote></details>

Yes, the call to `get_index()` is on the line before this line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaPuoYtwYjEToJ1qVFC:bko0qc3)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Ah, one use case where a user may want to lump multiple model instances into a single port is if the user loads a humanoid robot as multiple model instances. For example, a user may load a torso, head, left arm, right arm, left leg, and right leg as individual model instances, then want to treat the union of these six model instance ins as a single "humanoid" model instance. I see this use case as justifying the need to modify the API of `RigidBodyPlant` (or `MultiBodyPlant`) to allow users to say "create a new model instance by merging these six model instances". The user could then use the input / output ports of this newly created model instance. I will thus not pivot from the design I mentioned above.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaPz0rjAQJv2LEAbxiv:b-40y6z3)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sherm1 and @seancurtis-tri, I updated this PR per review comments. PTAL. The plant-centric actuator command input port is now only available when there is only one model instance present, and its accessor returns the same input port as the model instance-centric port. I believe this design is a lot cleaner per my posts above. Note that the plant-centric state output port is still exported regardless of the number of model instances present because it is used by Drake Visualizer. If we don't provide this output port, a Mux system will be needed to combine the states of each model instance.

---

Review status: 1 of 17 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 47 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaNHe-BBHW24-2DRbWj:-Ka_sI5TnyoJVBHlZIgK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f998efa1d6cac867756c41355a66a9292b70ecbe/drake/multibody/rigid_body_plant/rigid_body_plant.h#L47)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

It seems to me that between this "rule" and its dual above, it is worth noting that this exclusive connections are not enforced by the system and that breaking this "rule" will lead to...undefined behavior?  Some such thing.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOqRiyoEUu_NO9oa2I:-Ka_sEdolf-aw7MBoiOC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L31)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

If this is an actuator input it should have `actuator` in the name. There could be lots of different types of commands I think.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 33 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOmRaJH-UVQp9xkRO1:-Ka_sCdZtT1F7ZflLt8Z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L33)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This should not have both input ports if they can't both be used. Instead, specify on construction (or post-construction set methods) so that only one of the two input ports is created.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 71 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOpte1JGvPkKY4IyRT:-Ka_sLrH7E8NLRgZRvPw:b-bgsjpv) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, consider having both input and output obey the construction option. So you either get `all_actuators_in->RBPlant->all_states_out` or `model_instance_actuators_in->RBPlant->mode_instance_states_out`.
</blockquote></details>

The all_state_out port is needed by DrakeVisualizer. I thought keep it present would be OK to avoid needing a Mux. Is this OK? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-Ka_upuTRjPR5v0efzxp:b-vkvvl6)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please.

liangfok commented:
This PR passed all pre-merge CI checks and is thus ready for review.

---

Review status: 1 of 17 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KadwEef09f2yZ3IJxzC:b-5rq8uy)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Read through with some comments.

---

Reviewed 16 of 16 files at r7.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc, line 101 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah0nNXSdBGVBzH92UR:-Kah0nNXSdBGVBzH92US:b640lle) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc#L101)):*
> ```cpp
> 
>   // Fixes constant sources to all inputs.
>   const systems::RigidBodyPlant<T>& plant = visualized_plant_->plant();
> ```

BTW FTR, the semantics of this call: `visualized_plant_->plant()` seems a bit silly.  A plant's plant?  For my druthers, this kind of functionality is better served by having a `PlantVisualizer` that owns a plant.  The issue is that conceptually, we've got a subclass (i.e., a `RigidBodyPlant` with additional functionality) that has been implemented through composition.  The name suggests the class *is* a plant but it is, in fact, composed of one.  This is outside the scope of your PR, but I'd advocate revisiting the name of this and I figured I'd put that thought SOMEWHERE.  

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc, line 104 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah36sfLH21f0rDiFjR:-Kah36sfLH21f0rDiFjS:bj128h3) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc#L104)):*
> ```cpp
> 
>   for (int instance_id = RigidBodyTreeConstants::kFirstModelInstanceId;
>       instance_id < plant.get_num_model_instances(); ++instance_id) {
> ```

The semantics of `kFirstModelInstaceId` and `plant.get_num_model_instances()` don't line up.  If I'm told the start value (`s`) and the number of references (`n`), I'd expect to operate on the range `[s, s + n)`.  In this case, you're operating on the range `[s, n)`.  The constant is there to skip the world model instance, but `plant.get_num_model_instances()` reports a number *including* that instance (hence the apparent semantic inconsistency.) 

---

*[drake/multibody/rigid_body_tree.cc, line 2879 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah8TkVBkC45TrjRHbs:-Kah8TkVBkC45TrjRHbt:broc595) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_tree.cc#L2879)):*
> ```cpp
> 
> // TODO(liang.fok) Update this method implementation once the world is assigned
> // its own model instance ID (#3088). It should return num_model_instances_ - 1.
> ```

See previous note on semantics of this method -- particularly in light of your usage in `world_sim_diagram_factory.cc`.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 39 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHccsUB3uzPDaSdci:-KahHcctqizqLgVHoRvw:b4d8bqn) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L39)):*
> ```cpp
>   state_output_port_index_ =
>       this->DeclareOutputPort(kVectorValued, get_num_states()).get_index();
>   ExportModelInstanceCentricPorts();
> ```

Should this *not* be dependent on a boolean parameter to the constructor?

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 414 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahAgbqRXY0bl1R3NdP:-KahAgbqRXY0bl1R3NdQ:b-4e2ybc) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L414)):*
> ```cpp
> 
>   if (get_num_actuators() > 0) {
>     // The following code checks that all plant-centric actuator command input
> ```

This comment is misleading.  What this code does is *evaluate* the command input ports and throws an exception if one isn't connected.  When I first read this comment, my first thought was "should we be doing a structural integrity test every time we request derivatives?"  But when I looked at the code, I noticed the `this->EvalVectorInput` call which is more than just testing that things are plugged in.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 415 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahAWVPqZWUmWOOnYG8:-KahAWVPqZWUmWOOnYG9:b-cjr4vh) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L415)):*
> ```cpp
>   if (get_num_actuators() > 0) {
>     // The following code checks that all plant-centric actuator command input
>     // ports for are connected. It throws a runtime_error exception if this is
> ```

typo: "...ports *for* are connected."

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOqRiyoEUu_NO9oa2I:-KahBc3KNYlO53T0mGIH:bdukn5s) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L31)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'll go ahead and stick my oar in as well...  If I understand correctly, "actuator signals" are commands but not all commands are "actuator signals" (if that's the right term.)  Then `actuator_command_input_port` has an element of redundancy to it.  And when we have model-index-specific variants the names get monstrously long. Given the presence of the term "actuator", do we still need to include "command"?

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 76 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHjxa1D0WFmdZjjgc:-KahHjxa1D0WFmdZjjgd:b7kxs2e) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L76)):*
> ```objc
> ///   state vector for a particular model instance in the RigidBodyTree. This
> ///   method can only be called when this class is instantiated with constructor
> ///   parameter `export_model_instance_centric_ports` equal to `true`.
> ```

Am I missing this parameter in the constructor?

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 273 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahGgfK7Rt4uasdKo9O:-KahGgfK7Rt4uasdKo9P:bd0dko5) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L273)):*
> ```objc
> 
>   // Returns true if and only if the model instance with the provided
>   // `model_instance_id` has actuators. This is useful when trying to determine
> ```

use `@p model_instance_id` instead of \`model_instance_id\`

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 255 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHusq3LucoF6fe_cd:-KahHusq3LucoF6fe_ce:blp1436) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L255)):*
> ```cpp
>   ASSERT_EQ(1, kuka_plant_->get_num_model_instances());
> 
>   // TODO(liang.fok) Update this to be 1 once #3088 is resolved.
> ```

Could you not just assign it the value of `RigidBodyTreeConstants::kFirstModelInstanceId` (or whatever it was?)

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 295 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahIE-jSE2VlvtBxFZE:-KahIE-k48i42ANatYmC:bylpjux) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L295)):*
> ```cpp
>   //
>   //    (1) plant state
>   //    (2) model instance state
> ```

`(2) model instance state for tree containing a single model instance`

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_that_publishes_xdot_test.cc, line 1 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahIeZEJuEaf7Q4s-NQ:-KahIeZFmXS4DikIw2A7:b-tax7co) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_that_publishes_xdot_test.cc#L1)):*
> ```cpp
> #include "drake/multibody/rigid_body_plant/rigid_body_plant_that_publishes_xdot.h"
> ```

It's unclear that the changes to this file should be part of this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KahImIKgDK2vmUZ8MZe:bfhi5t4)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Cleared a few of my blockages and added a comment.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 273 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahGgfK7Rt4uasdKo9O:-KaiPwdxWbjDjHkIp1_V:b-im6ga6) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L273)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

use `@p model_instance_id` instead of \`model_instance_id\`
</blockquote></details>

BTW, doxygen treats
```
@p blah
`blah`
```
the same -- these are just font-changing commands. I prefer the backtick version myself because I find the `@p` construct distracting when I'm trying to read a comment while I have no trouble ignoring the backticks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaiQuYGTi_mTNPkx0hR:bvqvs69)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
And my riposte... :D

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 273 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahGgfK7Rt4uasdKo9O:-KaicluKrvGppmMayxM8:b-9pig7n) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L273)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, doxygen treats
```
@p blah
`blah`
```
the same -- these are just font-changing commands. I prefer the backtick version myself because I find the `@p` construct distracting when I'm trying to read a comment while I have no trouble ignoring the backticks.
</blockquote></details>

Ah, but in this case, file consistency wins.  :)  There are other instances of @p model_instance_id in this file already. (Otherwise, I agree.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KaicwAeOI2BBlKHZ-46:b-81ircs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 272 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kaj4jk4Ry9zW3B9-uZ9:-Kaj4jk4Ry9zW3B9-uZA:bajcuzj) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L272)):*
> ```objc
>   }
> 
>   // Returns true if and only if the model instance with the provided
> ```

This should be a `///` I believe!

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 273 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahGgfK7Rt4uasdKo9O:-Kaj4vOmk1jFaQR2i5N5:b29s2uv) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L273)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Ah, but in this case, file consistency wins.  :)  There are other instances of @p model_instance_id in this file already. (Otherwise, I agree.)
</blockquote></details>

Yeah, file consistency is good too. Not sure if it is overwhelming in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 9 of 16 files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc, line 101 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah0nNXSdBGVBzH92UR:-KamNQQ9UnMSQqQcNPPI:b86vyeo) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc#L101)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW FTR, the semantics of this call: `visualized_plant_->plant()` seems a bit silly.  A plant's plant?  For my druthers, this kind of functionality is better served by having a `PlantVisualizer` that owns a plant.  The issue is that conceptually, we've got a subclass (i.e., a `RigidBodyPlant` with additional functionality) that has been implemented through composition.  The name suggests the class *is* a plant but it is, in fact, composed of one.  This is outside the scope of your PR, but I'd advocate revisiting the name of this and I figured I'd put that thought SOMEWHERE.  
</blockquote></details>

Good point! Punting to #4813.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc, line 104 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah36sfLH21f0rDiFjR:-Kamiw4bgasZ0WYB0tax:bzek05s) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc#L104)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

The semantics of `kFirstModelInstaceId` and `plant.get_num_model_instances()` don't line up.  If I'm told the start value (`s`) and the number of references (`n`), I'd expect to operate on the range `[s, s + n)`.  In this case, you're operating on the range `[s, n)`.  The constant is there to skip the world model instance, but `plant.get_num_model_instances()` reports a number *including* that instance (hence the apparent semantic inconsistency.) 
</blockquote></details>

Done. I renamed the constant to be `kFirstNonWorldModelInstanceId` to indicate that is excludes the world and thus does not conflict with `plan.get_num_model_instances()`. Note that it is still zero until #3088 is resolved.

---

*[drake/multibody/rigid_body_tree.cc, line 2879 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah8TkVBkC45TrjRHbs:-KamlFo0Xe9sNQMZrnpu:bm5srzt) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_tree.cc#L2879)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

See previous note on semantics of this method -- particularly in light of your usage in `world_sim_diagram_factory.cc`.
</blockquote></details>

Done. I modified the constant variable's name to exclude the world. Until #3088 is resolved, confusion will likely persist, unfortunately.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 39 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHccsUB3uzPDaSdci:-KamlP_a8KpZwS0jJasz:b7i2o55) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L39)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Should this *not* be dependent on a boolean parameter to the constructor?
</blockquote></details>

This is correct; the model-instance-centric ports are now always exported.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 414 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahAgbqRXY0bl1R3NdP:-KammwIGI0CXhNJwNMbF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L414)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

This comment is misleading.  What this code does is *evaluate* the command input ports and throws an exception if one isn't connected.  When I first read this comment, my first thought was "should we be doing a structural integrity test every time we request derivatives?"  But when I looked at the code, I noticed the `this->EvalVectorInput` call which is more than just testing that things are plugged in.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 415 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahAWVPqZWUmWOOnYG8:-Kamn-4fHL4LpShgBFw3:b99x90p) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L415)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

typo: "...ports *for* are connected."
</blockquote></details>

Done (reworded per above comment).

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KaOqRiyoEUu_NO9oa2I:-KamnKR7aJqku869s0Ud:b-ur9g41) ([raw file](https://github.com/robotlocomotion/drake/blob/e728e4e36320e4a1f53c5a6feab53fc7b24e038e/drake/multibody/rigid_body_plant/rigid_body_plant.h#L31)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'll go ahead and stick my oar in as well...  If I understand correctly, "actuator signals" are commands but not all commands are "actuator signals" (if that's the right term.)  Then `actuator_command_input_port` has an element of redundancy to it.  And when we have model-index-specific variants the names get monstrously long. Given the presence of the term "actuator", do we still need to include "command"?
</blockquote></details>

I think it is OK to include "command" after "actuator" because otherwise an unsuspecting reader might be left wondering "what type of actuator data type should I be sending into this port?".

Looking more into the future, I can imagine other types of input information being conveyed to the actuator model. For example, maybe there will be an "actuator_desired_stiffness_input_port()", or "actuator_power_allotment_input_port()", etc., which should be distinguished from the actual command. 

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 76 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHjxa1D0WFmdZjjgc:-KamoP9JCIa_AYGe7BeM:baslr65) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L76)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Am I missing this parameter in the constructor?
</blockquote></details>

Oops! I forgot to delete that sentence. Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 272 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kaj4jk4Ry9zW3B9-uZ9:-KamoVQBc2aqkxNpve8V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L272)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This should be a `///` I believe!
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.h, line 273 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahGgfK7Rt4uasdKo9O:-KamoonMhtWFjVTouC7D:b1x9i5d) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.h#L273)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yeah, file consistency is good too. Not sure if it is overwhelming in this case.
</blockquote></details>

Done. Since @seancurtis-tri said "Otherwise, I agree", I decided to replace all instances of `@p foo` in this file.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 255 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHusq3LucoF6fe_cd:-Kamp88mMa4QYccwAypt:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L255)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Could you not just assign it the value of `RigidBodyTreeConstants::kFirstModelInstanceId` (or whatever it was?)
</blockquote></details>

Done. Thanks.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc, line 295 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahIE-jSE2VlvtBxFZE:-KampEPwTXLHEottabbF:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_test.cc#L295)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

`(2) model instance state for tree containing a single model instance`
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_plant/test/rigid_body_plant_that_publishes_xdot_test.cc, line 1 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahIeZEJuEaf7Q4s-NQ:-KamqDk6bJdL1kSLV_qj:b-9ks0s0) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/test/rigid_body_plant_that_publishes_xdot_test.cc#L1)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

It's unclear that the changes to this file should be part of this PR.
</blockquote></details>

Done: #4815.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KamqFcZkSH7xWkTSxhw:bblnx7i)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r8, 7 of 7 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc, line 104 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-Kah36sfLH21f0rDiFjR:-Kan0_gZdbudJoOxLwAl:bjumnnd) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_diagram_factory.cc#L104)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done. I renamed the constant to be `kFirstNonWorldModelInstanceId` to indicate that is excludes the world and thus does not conflict with `plan.get_num_model_instances()`. Note that it is still zero until #3088 is resolved.
</blockquote></details>

And given the semantics of `get_num_model_instances` are documented to exclude the world model, I'm content.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 39 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHccsUB3uzPDaSdci:-Kan0nWKKCIPhDrXHLQu:b-nt7uz5) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L39)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This is correct; the model-instance-centric ports are now always exported.
</blockquote></details>

I guess I missed the part of the discussion where they would be instantiated exclusively.  I certainly bought into the argument vis a vis *output* ports -- the plant-level output always being necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-Kan295a1HHVn1Ya2a34:bmxilld)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/rigid_body_plant.cc, line 39 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-KahHccsUB3uzPDaSdci:-Kan2_Qc8YwlJUdWALNb:biuu81i) ([raw file](https://github.com/robotlocomotion/drake/blob/368819db1d02096c56b3238675a81ce85af41679/drake/multibody/rigid_body_plant/rigid_body_plant.cc#L39)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I guess I missed the part of the discussion where they would be instantiated exclusively.  I certainly bought into the argument vis a vis *output* ports -- the plant-level output always being necessary.
</blockquote></details>

There is a plant-centric actuator command input port accessor called `RigidBodyPlant::actuator_command_input_port()` but (1) it can only be called when there is only one non-world model instance in the tree and (2) it returns the same port as `RigidBodyPlant::model_instance_actuator_command_input_port(RigidBodyTreeConstants::kFirstNonWorldModelInstanceId)`. Since the same physical port is returned, there is no longer a conflict bteween the two accessors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 16 files at r7, 1 of 2 files at r8, 6 of 7 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4766#-:-KasJ09-IHmpjk5U9LF7:bnfp4nl)*
<!-- Sent from Reviewable.io -->
